src/main.rs

```diff
@@ -64,42 +77,121 @@ fn main() -> Result<()> {
             let exit_code = keep_exec(backend, backend.shim(), binary, gdblisten)?;
             std::process::exit(exit_code);
         }
-        cli::Command::Run(run) => {
-            let modfile = File::open(run.module)?;
-            let open_fd = modfile.as_raw_fd();
-            // FIXME (v0.1.0 KEEP-CONFIG HACK): since we don't have any way to
-            // pass configuration or data into a keep yet, for v0.1.0 we've
-            // just hardcoded exec-wasmtime to assume the module is open for reading
-            // on FD3. That *should* always be the case here (since nothing
-            // above opens files or anything), but if that assumption is wrong
-            // then things will break mysteriously later on. So this assert
-            // is just here to make them break earlier, and with less mystery.
-            assert!(open_fd == 3, "module got unexpected fd {}", open_fd);
-
-            let configfile = match run.exec.wasmcfgfile.as_ref() {
-                Some(name) => {
-                    let file = File::open(name)?;
-                    let fd = file.as_raw_fd();
-                    assert!(fd == 4, "config got unexpected fd {}", fd);
-
-                    Some(file)
+        cli::Command::Run(run::Options {
+            backend,
+            package,
+            steward,
+            #[cfg(feature = "gdb")]
+            gdblisten,
+        }) => {
+            let backend = backend.pick()?;
+            let exec = EXECS
+                .deref()
+                .iter()
+                .find(|w| w.with_backend(backend))
+                .ok_or_else(|| anyhow!("no supported exec found"))
+                .map(|b| b.exec())?;
+
+            let (exec_sock, mut host_sock) =
+                UnixStream::pair().context("failed to create a Unix socket pair")?;
+            assert_eq!(exec_sock.as_raw_fd(), 3);
+            assert_eq!(host_sock.as_raw_fd(), 4);
+
+            let (package, wasm, conf) = match package.scheme() {
+                "file" => {
+                    let path = package.to_file_path().map_err(|()| {
+                        anyhow!("failed to parse file path from URL `{}`", package)
+                    })?;
+                    match fs::metadata(&path).with_context(|| {
+                        format!(
+                            "failed to get information about `{}`",
+                            path.to_string_lossy(),
+                        )
+                    })? {
```

Comment:
I find it difficult to read the match with such a long expression here, can we split this out into its own binding? This match is already not really using patterns, so then we can just have `if md.is_file() { ... } else if md.is_dir() { ... } else {`